File: delta_star.cpp
#include<iostream>
#include<cmath>
using namespace std;

// Delta-Star transform
// used this for Tutorial 7 assignment
// R1=A-B, R2=B-C, R3=C-A

int main()
{
	for(;;)
	{
		double r1,r2,r3;
		cout<<"Enter R1 (A-B): ";
		cin>>r1;
		cout<<"Enter R2 (B-C): ";
		cin>>r2;
		cout<<"Enter R3 (C-A): ";
		cin>>r3;
		
		if(r1<=0||r2<=0||r3<=0)
		{
			cout<<"invalid input try again"<<endl;
			cout<<endl;
			system("pause");
			cout<<endl;
			continue;
		}
		
		double s = r1+r2+r3;
		double ra = r2*r3/s;
		double rb = r1*r3/s;
		double rc = r1*r2/s;
		
		cout<<endl;
		cout<<"Star equivalent:"<<endl;
		cout<<"  Ra (at A) = "<<ra<<" ohm"<<endl;
		cout<<"  Rb (at B) = "<<rb<<" ohm"<<endl;
		cout<<"  Rc (at C) = "<<rc<<" ohm"<<endl;
		
		// verify
		cout<<endl;
		cout<<"Verification:"<<endl;
		double d_ab = r1*(r2+r3)/s;
		double s_ab = ra+rb;
		cout<<"  A-B: Delta="<<d_ab<<" Star="<<s_ab;
		if(abs(d_ab-s_ab)<0.01) cout<<" OK";
		else cout<<" FAIL";
		cout<<endl;
		
		// reverse: star to delta
		double p = ra*rb+rb*rc+ra*rc;
		double r1b = p/rc;
		double r2b = p/rb;
		double r3b = p/ra;
		cout<<endl;
		cout<<"Reverse (Star->Delta):"<<endl;
		cout<<"  R1 = "<<r1b<<" ohm"<<endl;
		cout<<"  R2 = "<<r2b<<" ohm"<<endl;
		cout<<"  R3 = "<<r3b<<" ohm"<<endl;
		
		cout<<endl;
		system("pause");
		cout<<endl;
	}
	return 0;
}

File: delta_star.py
"""
Delta-Star (triangle <-> Y) transformation.
Used this for Tutorial 7 assignment.
"""


def delta_to_star(r1, r2, r3):
    # R1=A-B, R2=B-C, R3=C-A
    # Ra=at A, Rb=at B, Rc=at C
    s = r1 + r2 + r3
    ra = r2 * r3 / s
    rb = r1 * r3 / s
    rc = r1 * r2 / s
    return (ra, rb, rc)


def star_to_delta(ra, rb, rc):
    p = ra * rb + rb * rc + ra * rc
    r1 = p / rc
    r2 = p / rb
    r3 = p / ra
    return (r1, r2, r3)


def verify(r1, r2, r3, ra, rb, rc):
    s = r1 + r2 + r3
    # check all 3 terminal pairs
    d_ab = r1 * (r2 + r3) / s
    d_bc = r2 * (r1 + r3) / s
    d_ca = r3 * (r1 + r2) / s
    ok = (abs(d_ab - (ra + rb)) < 0.01 * d_ab and
          abs(d_bc - (rb + rc)) < 0.01 * d_bc and
          abs(d_ca - (rc + ra)) < 0.01 * d_ca)
    return ok


if __name__ == "__main__":
    # Tutorial 7 values
    r1, r2, r3 = 680, 1000, 1200
    print(f"Delta: R1={r1}, R2={r2}, R3={r3}")

    ra, rb, rc = delta_to_star(r1, r2, r3)
    print(f"Star: Ra={ra:.1f}, Rb={rb:.1f}, Rc={rc:.1f}")

    if verify(r1, r2, r3, ra, rb, rc):
        print("verification OK")
    else:
        print("verification FAILED")

    # reverse check
    r1b, r2b, r3b = star_to_delta(ra, rb, rc)
    print(f"\nReverse: R1={r1b:.1f}, R2={r2b:.1f}, R3={r3b:.1f}")
